README.md
# WooCommerce Product Bundles - Top Add to Cart Button

### What's This?

Mini-extension for [WooCommerce Product Bundles](https://woocommerce.com/products/product-bundles/) that adds an extra add-to-cart button section at the top of Product Bundle pages.

![Additional add-to-cart section added before a Product Bundle form.](https://user-images.githubusercontent.com/1783726/29972494-ef0fb1a2-8f34-11e7-8d98-9ebea97026d2.png)

### Important

Requires WooCommerce Product Bundles 5.5+.


changelog.txt
*** WooCommerce Product Bundles - Top Add to Cart Button ***

2018.09.25 - version 1.0.3
* Tweak - Declare WooCommerce 3.5 support.

2017.11.11 - version 1.0.2
* Tweak - Updated readme.md file.

2017.11.05 - version 1.0.1
* Tweak - Added readme.txt file.
* Tweak - Added WC version support headers.

2017.09.01 - version 1.0.0
* Initial Release.


Gruntfile.js
/* jshint node:true */
module.exports = function( grunt ) {
	'use strict';

	grunt.initConfig({

		// Setting folder templates.
		dirs: {
			css: 'assets/css',
			js: 'assets/js'
		},

		// JavaScript linting with JSHint.
		jshint: {
			options: {
				'force': true,
				'boss': true,
				'curly': true,
				'eqeqeq': false,
				'eqnull': true,
				'es3': false,
				'expr': false,
				'immed': true,
				'noarg': true,
				'onevar': true,
				'quotmark': 'single',
				'trailing': true,
				'undef': true,
				'unused': true,
				'sub': false,
				'browser': true,
				'maxerr': 1000,
				globals: {
					'jQuery': false,
					'$': false,
					'Backbone': false,
					'_': false,
					'wp': false,
					'wc_bundle_params': false,
					'woocommerce_params': false,
					'woocommerce_admin_meta_boxes': true,
					'woocommerce_writepanel_params': false,
					'wc_composite_admin_params': false,
					'wc_bundle_admin_params': false,
					'wc_enhanced_select_params': false,
					'PhotoSwipe': true,
					'PhotoSwipeUI_Default': true
				},
			},
			all: [
				'!Gruntfile.js',
				'<%= dirs.js %>/*.js',
				'!<%= dirs.js %>/*.min.js'
			]
		},
		// Minify .js files.
		uglify: {
			options: {
				preserveComments: 'some'
			},
			jsfiles: {
				files: [{
					expand: true,
					cwd: '<%= dirs.js %>',
					src: [
						'*.js',
						'!*.min.js'
					],
					dest: '<%= dirs.js %>',
					ext: '.min.js'
				}]
			}
		},

		// Watch changes for assets.
		watch: {
			js: {
				files: [
					'<%= dirs.js %>/*js'
				],
				tasks: ['uglify']
			}
		},

		// Generate POT files.
		makepot: {
			options: {
				type: 'wp-plugin',
				domainPath: 'languages',
				potHeaders: {
					'report-msgid-bugs-to': 'support@somewherewarm.gr',
				}
			},
			go: {
				options: {
					potFilename: 'woocommerce-product-bundles-top-add-to-cart-button.pot',
					exclude: [
						'languages/.*',
						'assets/.*',
						'node-modules/.*',
					]
				}
			}
		},

		// Check textdomain errors.
		checktextdomain: {
			options:{
				text_domain: 'woocommerce-product-bundles-top-add-to-cart-button',
				keywords: [
					'__:1,2d',
					'_e:1,2d',
					'_x:1,2c,3d',
					'esc_html__:1,2d',
					'esc_html_e:1,2d',
					'esc_html_x:1,2c,3d',
					'esc_attr__:1,2d',
					'esc_attr_e:1,2d',
					'esc_attr_x:1,2c,3d',
					'_ex:1,2c,3d',
					'_n:1,2,4d',
					'_nx:1,2,4c,5d',
					'_n_noop:1,2,3d',
					'_nx_noop:1,2,3c,4d'
				]
			},
			files: {
				src:  [
					'**/*.php', // Include all files.
					'!node_modules/**' // Exclude node_modules/ dir.
				],
				expand: true
			}
		}
	});

	// Load NPM tasks to be used here.
	grunt.loadNpmTasks( 'grunt-contrib-jshint' );
	grunt.loadNpmTasks( 'grunt-contrib-uglify' );
	grunt.loadNpmTasks( 'grunt-contrib-watch' );
	grunt.loadNpmTasks( 'grunt-wp-i18n' );
	grunt.loadNpmTasks( 'grunt-checktextdomain' );

	// Register tasks.
	grunt.registerTask( 'dev', [
		'checktextdomain',
		'uglify',
	]);

	grunt.registerTask( 'default', [
		'dev',
		'makepot'
	]);

	grunt.registerTask( 'domain', [
		'checktextdomain'
	]);
};


assets/js/wc-pb-top-add-to-cart-button.js
;( function ( $, window, document, undefined ) {

	var PB_Integration = function( $top, bundle ) {

		this.$qty       = false;
		this.qty_synced = false;

		var self = this;

		// Init.
		this.integrate = function() {

			var $bundle_button       = $top.find( '.bundle_button' ),
				$bundle_error        = $top.find( '.bundle_error' ),
				$bundle_price        = $top.find( '.bundle_price' ),
				$bundle_availability = $top.find( '.bundle_availability' );

			if ( $bundle_button.length > 0 ) {
				$bundle_button.append( bundle.$bundle_button.html() );
				bundle.$bundle_button.push( $bundle_button.get( 0 ) );
				self.$qty = $top.find( 'input.qty' );
			}

			if ( $bundle_error.length > 0 ) {
				$bundle_error.append( bundle.$bundle_error.html() );
				bundle.$bundle_error.push( $bundle_error.get( 0 ) );
				bundle.$bundle_error_content.push( $bundle_error.find( 'ul.msg' ).get( 0 ) );
			}

			if ( $bundle_price.length > 0 ) {
				$bundle_price.append( bundle.$bundle_price.html() );
				bundle.$bundle_price.push( $bundle_price.get( 0 ) );
			}

			if ( $bundle_availability.length > 0 ) {
				$bundle_availability.append( bundle.$bundle_availability.html() );
				bundle.$bundle_availability.push( $bundle_availability.get( 0 ) );
			}
		};

		this.add_hooks = function() {

			// Relay button click to actual form button.
			$top.on( 'submit', function() {
				bundle.$bundle_button.find( '.bundle_add_to_cart_button' ).trigger( 'click' );
				return false;
			} );

			// Keep bundle quantities in sync: Top quantity changed.
			if ( self.$qty ) {
				self.$qty.on( 'input change', function() {

					if ( ! self.qty_synced ) {
						self.qty_synced = true;
						bundle.$bundle_quantity.val( self.$qty.val() ).change();
						self.qty_synced = false;
					}
				} );
			}

			// Keep bundle quantities in sync: Bottom quantity changed.
			bundle.$bundle_quantity.on( 'input change', function() {
				self.$qty.val( bundle.$bundle_quantity.val() ).change();
			} );
		};

		// Lights on.
		this.integrate();
		this.add_hooks();
	};

	// Hook into Bundles.
	$( '.bundle_form_top' ).each( function() {
		var $top = $( this ),
			$bundle_data = $( this ).closest( '.product' ).find( '.bundle_form .bundle_data' );

		if ( $bundle_data.length > 0 ) {
			$bundle_data.on( 'woocommerce-product-bundle-initializing', function( event, bundle ) {
				new PB_Integration( $top, bundle );
			} );
		}
	} );

} )( jQuery, window, document );


woocommerce-product-bundles-top-add-to-cart-button.php
<?php
/*
* Plugin Name: WooCommerce Product Bundles - Top Add to Cart Button
* Plugin URI: http://woocommerce.com/products/product-bundles/
* Description: Adds an add-to-cart button section at the top of Product Bundle pages.
* Version: 1.0.3
* Author: SomewhereWarm
* Author URI: http://somewherewarm.gr/
*
* Text Domain: woocommerce-product-bundles-top-add-to-cart-button
* Domain Path: /languages/
*
* Requires at least: 4.1
* Tested up to: 4.9
*
* WC requires at least: 3.0
* WC tested up to: 3.5
*
* Copyright: © 2017 SomewhereWarm SMPC.
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

class WC_PB_Top_Add_To_Cart {

	/**
	 * Version.
	 * @var string
	 */
	public static $version = '1.0.3';

	/**
	 * Required PB version.
	 * @var string
	 */
	public static $req_pb_version = '5.5';

	/**
	 * Plugin URL.
	 */
	public static function plugin_url() {
		return plugins_url( basename( plugin_dir_path(__FILE__) ), basename( __FILE__ ) );
	}

	/**
	 * Plugin path.
	 */
	public static function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	/**
	 * Entry point.
	 */
	public static function init() {
		add_action( 'plugins_loaded', array( __CLASS__, 'load_plugin' ) );
	}

	/**
	 * Lights on.
	 */
	public static function load_plugin() {

		if ( ! function_exists( 'WC_PB' ) || version_compare( WC_PB()->version, self::$req_pb_version ) < 0 ) {
			add_action( 'admin_notices', array( __CLASS__, 'version_notice' ) );
			return false;
		}

		// Localize plugin.
		add_action( 'init', array( __CLASS__, 'localize_plugin' ) );

		// Hook extra add-to-cart section.
		add_action( 'woocommerce_single_product_summary', array( __CLASS__, 'top_add_to_cart' ), 25 );

		// Enqueue scripts and styles.
		add_action( 'wp_enqueue_scripts', array( __CLASS__, 'script' ) );
	}

	/**
	 * Load textdomain.
	 *
	 * @return void
	 */
	public static function localize_plugin() {
		load_plugin_textdomain( 'woocommerce-product-bundles-top-add-to-cart-button', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
	}

	/**
	 * PB version check notice.
	 */
	public static function version_notice() {
	    echo '<div class="error"><p>' . sprintf( __( '<strong>WooCommerce Product Bundles &ndash; Top Add To Cart Button</strong> requires Product Bundles <strong>%s</strong> or higher.', 'woocommerce-product-bundles-top-add-to-cart-button' ), self::$req_pb_version ) . '</p></div>';
	}

	/**
	 * Show extra add-to-cart button section.
	 */
	public static function top_add_to_cart() {

		global $product;

		if ( is_a( $product, 'WC_Product_Bundle' ) ) {

			$elements      = apply_filters( 'woocommerce_bundles_top_add_to_cart_elements', array( 'error', 'availability', 'button' ), $product );
			$elements_html = '';

			foreach ( $elements as $el ) {
				$elements_html .= '<div class="bundle_' . $el . '"' . ( 'error' === $el ? ' style="display:none"' : '' ) . '></div>';
			}

			if ( ! empty( $elements_html ) ) {
				echo '<form method="post" class="cart bundle_form bundle_form_top"><div class="bundle_wrap">' . $elements_html . '</div></form>';
			}
		}
	}

	/**
	 * Script.
	 */
	public static function script() {

		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';

		wp_register_script( 'wc-pb-top-add-to-cart-button', self::plugin_url() . '/assets/js/wc-pb-top-add-to-cart-button' . $suffix . '.js', array( 'wc-add-to-cart-bundle' ), self::$version, true );
		wp_enqueue_script( 'wc-pb-top-add-to-cart-button' );
	}
}

WC_PB_Top_Add_To_Cart::init();
